### PR TITLE
Split TopLevelExportDefs away from LinkedClass

### PR DESCRIPTION
**Pull request by gzm0**
Side-effects for TopLevelMethodExportDefs in IRChecker:
- Allow them on all class kinds.
- Allow names containing `__` for non JS classes.
- Disallow references to JS class captures (this was a bug).

### CODE REVIEW

**Comment by gzm0** [comment]
Split away from #4111. To be merged now, or not :)

File: linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala [hunk 1]
```diff
@@ -42,8 +43,6 @@ object Infos {
       val methods: List[MethodInfo],
       val jsNativeMembers: Map[MethodName, JSNativeLoadSpec],
       val exportedMembers: List[ReachabilityInfo],
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala:45]: The CI says:
```suggestion
      val exportedMembers: List[ReachabilityInfo]
```

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/IRLoader.scala [hunk 2]
```diff
@@ -91,20 +104,25 @@ final class IRLoader extends Analyzer.InputProvider with MethodSynthesizer.Input
...
+  }
+
   def loadClassDefAndVersion(irFile: IRFileImpl)(
       implicit ec: ExecutionContext): Future[(ClassDef, Option[String])] = {
     update(irFile).map(s => (s._1, version))
   }
 
   private def update(irFile: IRFileImpl)(
-      implicit ec: ExecutionContext): Future[(ClassDef, Infos.ClassInfo)] = synchronized {
+      implicit ec: ExecutionContext): Future[(ClassDef, Infos.ClassInfo, List[Infos.TopLevelExportInfo])] = synchronized {
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/IRLoader.scala:125]: The CI says that this line is too long (more than 120 characters). Perhaps move the `synchronized` inside?

Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/IRLoader.scala]: I took the chance to simplify this a bit. PTAL

**Pull request merged**